File: Codefun/dewu/08233.py
import sys


class Node:
    def __init__(self, u, v, w):
        self.u = u
        self.v = v
        self.w = w

    def __lt__(self, other):
        return self.w > other.w


def find(x):
    return x if fa[x] == x else find(fa[x])


def merge(x, y):
    fx = find(x)
    fy = find(y)
    if fx != fy:
        fa[fx] = fy


if __name__ == "__main__":
    n, m = map(int, sys.stdin.readline().split())
    e = [Node(0, 0, 0) for i in range(0, m)]
    for i in range(0, m):
        u, v, w = map(int, sys.stdin.readline().split())
        e[i] = Node(u, v, w)
    fa = [i for i in range(0, m + 1)]
    e.sort()
    cnt = 1
    ans = 0
    for i in range(0, m):
        u = e[i].u
        v = e[i].v
        w = e[i].w
        if find(u) != find(v):
            merge(u, v)
            cnt += 1
            ans += w
            if cnt == n:
                print(ans)
                exit()
    print("No solution.")

File: theory/Codefun/dewu/08233.cc
#include <bits/stdc++.h>
using namespace std;

struct node
{
    int u, v, w;
    // 端点 u , v , 权值 w
    bool operator<(const node &a) const
    {
        return w > a.w;
    }
};
int fa[1010]; // 并查集
int find(int x)
{
    return x == fa[x] ? x : fa[x] = find(fa[x]);
}
void merge(int x, int y)
{
    int fx = find(x), fy = find(y);
    if (fx != fy)
        fa[fx] = fy;
}

int main()
{
    ios::sync_with_stdio(false);

    int n, m;
    cin >> n >> m;
    vector<node> e(m + 1);
    for (int i = 1; i <= m; i++)
    {
        cin >> e[i].u >> e[i].v >> e[i].w;
    }
    for (int i = 1; i <= m; i++)
    {
        fa[i] = i;
    }
    sort(e.begin() + 1, e.end());
    int cnt = 1;
    long long ans = 0;
    for (int i = 1; i <= m; i++)
    {
        auto [u, v, w] = e[i];
        if (find(u) != find(v))
        {
            merge(u, v);
            cnt++;
            ans += w;
            if (cnt == n)
            {
                cout << ans << endl;
                return 0;
            }
        }
    }
    cout << "No solution."<< endl;
}

